rustls/src/ticketer.rs

```diff
@@ -54,83 +52,49 @@ impl TicketSwitcher {
-        //
-        // There are a few things to note here. First, we don't check whether
-        // a new switch might be needed in step 4, even though, due to locking
-        // and entropy collection, significant amounts of time may have passed.
-        // This is to guarantee that the thread doing the switch will eventually
-        // make progress.
-        //
-        // Second, because next may be None, step 2 can fail. In that case
-        // we enter a recovery mode where we generate 2 new ticketers, one for
-        // next and one for the current ticketer. We then take the mutex a
-        // second time and redo the time check to see if a switch is still
-        // necessary.
-        //
-        // This somewhat convoluted approach ensures good availability of the
-        // mutex, by ensuring that the state is usable and the mutex not held
-        // during generation. It also ensures that, so long as the inner
-        // ticketer never generates panics during encryption/decryption,
-        // we are guaranteed to never panic when holding the mutex.
-
+    /// For efficiency, this is also responsible for locking the state rwlock
+    /// and returning it for read.
+    pub(crate) fn maybe_roll(
+        &self,
+        now: UnixTime,
+    ) -> Option<RwLockReadGuard<'_, TicketSwitcherState>> {
         let now = now.as_secs();
-        let mut are_recovering = false; // Are we recovering from previous failure?
-        {
-            // Scope the mutex so we only take it for as long as needed
-            let mut state = self.state.lock()?;
 
-            // Fast path in case we do not need to switch to the next ticketer yet
-            if now <= state.next_switch_time {
-                return Some(state);
-            }
+        // Fast, common, & read-only path in case we do not need to switch
+        // to the next ticketer yet
+        {
+            let read = self.state.read().ok()?;
 
-            // Make the switch, or mark for recovery if not possible
-            if let Some(next) = state.next.take() {
-                state.previous = Some(mem::replace(&mut state.current, next));
-                state.next_switch_time = now.saturating_add(u64::from(self.lifetime));
-            } else {
-                are_recovering = true;
+            if now <= read.next_switch_time {
+                return Some(read);
             }
         }
 
-        // We always need a next, so generate it now
+        // We need to switch ticketers, and make a new one.
+        // Generate a potential "next" ticketer outside the lock.
         let next = (self.generator)().ok()?;
-        if !are_recovering {
-            // Normal path, generate new next and place it in the state
-            let mut state = self.state.lock()?;
-            state.next = Some(next);
-            Some(state)
+
+        let mut write = self.state.write().ok()?;
+
+        if now > write.next_switch_time {
```

Comment:
I think this might be easier to follow if you do an early return for the "another thread beat us to it" case, at the cost of a little code duplication?